README.md
# Kodi Hyperion Addon

Test install guide: place the content of this repository into ```~/.kodi/addons/script.hyperion/```

German Support Thread: http://powerpi.de/forum/viewtopic.php?t=190


changelog.txt
version 0.0.5
 - switch from hyperion remote-binary to json interface

version 0.0.4
 - switch additional executable

version 0.0.3
 - small bug fixes
 - code improvements

version 0.0.2
 - working service interface
 - enhanced json rpc interface

version 0.0.1
  - initial release
  - working json rpc interface for color/effect/clear commands


service.py
import xbmc, xbmcaddon, xbmcgui
from resources.lib import hyperion

__addon__ = xbmcaddon.Addon()
__title__ = __addon__.getAddonInfo('name')
__icon__ = __addon__.getAddonInfo('icon')
__language__ = __addon__.getLocalizedString
__hyperion__ = None


def log(txt):
    message = '%s: %s' % (__title__, txt.encode('ascii', 'ignore'))
    xbmc.log(msg=message, level=xbmc.LOGNOTICE)


class Main:
    def __init__(self):
        hyperion_host = __addon__.getSetting('hyperion_host')
        hyperion_port = __addon__.getSetting('hyperion_port')
        priority = __addon__.getSetting('priority')
        global __hyperion__

        if not hyperion_host or not hyperion_port:
            log('Service: ' + __language__(20000))
            xbmcgui.Dialog().notification(__title__, __language__(20000), __icon__, 5000)
            return

        __hyperion__ = hyperion.Remote(hyperion_host=hyperion_host, hyperion_port=hyperion_port, priority=priority)

        self.Player = MyPlayer()
        self.Monitor = MyMonitor()
        self._daemon()

    def _daemon(self):
        self.Monitor.onStart()
        while not self.Monitor.abortRequested():
            if self.Monitor.waitForAbort(10):
                self.Monitor.onShutdown()
                break


class MyMonitor(xbmc.Monitor):
    def __init__(self):
        xbmc.Monitor.__init__(self)

    def onStart(self):
        log('Monitor: onStart')
        if __addon__.getSetting('clear_on_start') == 'true':
            __hyperion__.clearAll()

        if __addon__.getSetting('effect_on_kodi_startup'):
            __hyperion__.effect(__addon__.getSetting('effect_on_kodi_startup'))

    def onShutdown(self):
        log('Monitor: onShutdown')
        if __addon__.getSetting('off_on_shutdown') == 'true':
            __hyperion__.color('black')

    def onScreensaverActivated(self):
        log('Monitor: onScreensaverActivated')
        if __addon__.getSetting('off_on_screensaver_activated') == 'true':
            __hyperion__.color('black')

    def onScreensaverDeactivated(self):
        log('Monitor: onScreensaverDeactivated')
        if __addon__.getSetting('effect_on_screensaver_deactived'):
            __hyperion__.effect(__addon__.getSetting('effect_on_screensaver_deactived'))


class MyPlayer(xbmc.Player):
    def __init__(self):
        xbmc.Player.__init__(self)

    def onPlayBackStarted(self):
        log('Player: onPlayBackStarted')
        if __addon__.getSetting('clear_on_video_playback') and xbmc.Player().isPlayingVideo():
            __hyperion__.clearAll()

        if __addon__.getSetting('effect_on_playback_audio') and xbmc.Player().isPlayingAudio():
            __hyperion__.effect(__addon__.getSetting('effect_on_playback_audio'))

    def onPlayBackResumed(self):
        if __addon__.getSetting('clear_on_video_playback') == 'true' and xbmc.Player().isPlayingVideo():
            __hyperion__.clearAll()

    def onPlayBackPaused(self):
        if __addon__.getSetting('effect_on_playback_paused') and xbmc.Player().isPlayingVideo():
            __hyperion__.effect(__addon__.getSetting('effect_on_playback_paused'))

    def onPlayBackStopped(self):
        if __addon__.getSetting('effect_on_playback_stopped'):
            __hyperion__.effect(__addon__.getSetting('effect_on_playback_stopped'))

    def onPlayBackEnded(self):
        self.onPlayBackStopped()


if __name__ == '__main__':
    if __addon__.getSetting('autostart') == 'true':
        log('Service: autostart execute')
        Main()


autostart.py
import xbmc
import xbmcaddon

settings = xbmcaddon.Addon(id='script.hyperion')

autostart = settings.getSetting('autostart')


class HyperionWatch(xbmc.Player):
    def __init__(self):
        xbmc.Player.__init__(self)

    def onPlayBackStarted(self):
        if xbmc.Player().isPlayingVideo():
            # TODO
            xbmc.executebuiltin('RunScript(/usr/local/bin/hyperion-clear.py)')  # clear all effect

        if xbmc.Player().isPlayingAudio():
            # TODO
            xbmc.executebuiltin('RunScript(/usr/local/bin/hyperion-effect-rainbow-swirl.py)')  # start effect

    def onPlayBackEnded(self):
        # TODO
        xbmc.executebuiltin('RunScript(/usr/local/bin/hyperion-effect-rainbow-swirl.py)')  # start effect

    def onPlayBackStopped(self):
        # TODO
        xbmc.executebuiltin('RunScript(/usr/local/bin/hyperion-effect-rainbow-swirl.py)')  # start effect

    def onPlayBackPaused(self):
        if xbmc.Player().isPlayingVideo():
            # TODO
            xbmc.executebuiltin('RunScript(/usr/local/bin/hyperion-effect-rainbow-swirl.py)')  # start effect

    def onPlayBackResumed(self):
        if xbmc.Player().isPlayingVideo():
            # TODO
            xbmc.executebuiltin('RunScript(/usr/local/bin/hyperion-clear.py)')  # clear all effect


if (autostart == "true"):
    HyperionWatch()

    # TODO ??
    # while(1):
    #     if xbmc.Player().isPlayingVideo():
    #         VIDEO = 1
    #     else:
    #         VIDEO = 0
    #
    #     xbmc.sleep(3000)


resources/lib/hyperion.py
import json
import socket
import time
import xbmc

from resources.lib import webcolors


def log(txt):
    message = '%s: %s' % ('Hyperion.Remote', txt.encode('ascii', 'ignore'))
    xbmc.log(msg=message, level=xbmc.LOGNOTICE)


class Remote:
    def __init__(self, hyperion_host='127.0.0.1', hyperion_port='19444', priority='0'):
        self.hyperion_host = hyperion_host
        self.hyperion_port = hyperion_port
        self.priority = priority

        try:
            import StorageServer
        except:
            from resources.lib import storageserverdummy as StorageServer

        self.cache = StorageServer.StorageServer('hyperion', 8544)

    def color(self, color, priority=None):

        payload = {'command': 'color', 'color': webcolors.name_to_rgb(color), 'priority': priority}

        if color == 'black':
            self.setState('off')
        else:
            self.setState('on')

        return self.run(payload)

    def effect(self, effect, priority=None):
        payload = {'command': 'effect', 'effect': {'name': effect}, 'priority': priority}

        self.setState('on')

        return self.run(payload)

    def clear(self, priority=None):
        payload = {'command': 'clear', 'priority': priority}

        self.setState('on')

        return self.run(payload)

    def clearAll(self):
        payload = {'command': 'clearall'}
        self.setState('on')

        return self.run(payload)

    def serverinfo(self):
        payload = {'command': 'serverinfo'}

        return self.run(payload)

    def run(self, payload):

        if not payload:
            return False

        if 'priority' in payload and payload['priority'] is None:
            payload['priority'] = int(self.priority)

        data = json.dumps(payload) + '\n'

        try:
            ret = self.nc(data)
        except Exception as e:
            log(str(data))
            log(str(e))
            return False

        return ret

    def setState(self, state):
        self.cache.set('state', state)

    def getState(self):
        return self.cache.get('state')

    def nc(self, data):
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

            s.connect((self.hyperion_host, int(self.hyperion_port)))

            s.sendall(data)

            data = self._recv(s)

            s.close()

        except socket.error as e:
            raise Exception(e)

        return json.loads(data)

    def _recv(self, s, timeout=0.1):
        # make socket non blocking
        s.setblocking(0)

        # total data partwise in an array
        total_data = []
        data = ''

        # beginning time
        begin = time.time()
        while 1:
            # if you got some data, then break after timeout
            if total_data and time.time() - begin > timeout:
                break

            # if you got no data at all, wait a little longer, twice the timeout
            elif time.time() - begin > timeout * 2:
                break

            # recv something
            try:
                data = s.recv(8192)
                if data:
                    total_data.append(data)
                    # change the beginning time for measurement
                    begin = time.time()
                else:
                    # sleep for sometime to indicate a gap
                    time.sleep(0.1)
            except:
                pass

        # join all parts to make final string
        return ''.join(total_data)


default.py
import sys
import subprocess
import xbmc
import xbmcaddon
import xbmcgui

from urlparse import parse_qsl
from resources.lib import hyperion

__addon__ = xbmcaddon.Addon()
__title__ = __addon__.getAddonInfo('name')
__icon__ = __addon__.getAddonInfo('icon')
__language__ = __addon__.getLocalizedString
__hyperion__ = None


def log(txt):
    message = '%s: %s' % (__title__, txt.encode('ascii', 'ignore'))
    xbmc.log(msg=message, level=xbmc.LOGNOTICE)


class Main:
    def __init__(self):
        hyperion_host = __addon__.getSetting('hyperion_host')
        hyperion_port = __addon__.getSetting('hyperion_port')
        priority = __addon__.getSetting('priority')

        global __hyperion__

        if not hyperion_host or not hyperion_port:
            print 'Hyperion.Script: ' + __language__(20000)
            xbmcgui.Dialog().notification(__title__, __language__(20000), __icon__, 5000)
            return

        __hyperion__ = hyperion.Remote(hyperion_host=hyperion_host, hyperion_port=hyperion_port, priority=priority)

        self.params = dict(parse_qsl('&'.join(sys.argv)))

        if not self.params:
            self._gui()
            return

        command = self.params.get('command', None)

        log('Script: command: %s' % command)

        if command is None:
            log('Script: no command given?!')

        elif command == 'effect':
            self._effect()

        elif command == 'color':
            self._color()

        elif command == 'clear':
            self._clear()

        elif command == 'clearall':
            self._clearAll()

        elif command == 'switch':
            self._switch()

    def _gui(self):
        # TODO cache serverinfo temporary?
        info = __hyperion__.serverinfo()

        if type(info) == type(True) or 'info' not in info:
            xbmcgui.Dialog().notification(__title__, __language__(20002), __icon__, 3000)
            return False

        if 'effects' not in info['info'] or not info['info']:
            xbmcgui.Dialog().notification(__title__, __language__(20003), __icon__, 3000)
            return False

        options = ['clear', 'clearAll', 'switch', '']

        for effect in info['info']['effects']:
            options.append(effect['name'])

        ret = xbmcgui.Dialog().select(__title__, options)

        if ret == -1:
            pass

        elif ret == 0:
            self._clear()

        elif ret == 1:
            self._clearAll()

        elif ret == 2:
            # if xbmcgui.Dialog().yesno(__title__, 'switch?'):
            self._switch()

        elif ret == 3:
            pass

        elif ret >= 4:
            self.params.setdefault('effect', options[ret])
            self._effect()

    def _effect(self):
        effect = self.params.get('effect', None)
        priority = self.params.get('priority', None)

        if not effect:
            return

        xbmcgui.Dialog().notification(__title__, effect, __icon__, 3000)
        __hyperion__.effect(effect, priority)

    def _color(self):
        color = self.params.get('color', None)
        priority = self.params.get('priority', None)

        if not color:
            return

        try:
            __hyperion__.color(color, priority)
        except ValueError as e:
            color = e.message

        xbmcgui.Dialog().notification(__title__, color, __icon__, 3000)

    def _clear(self):
        priority = self.params.get('priority', None)

        xbmcgui.Dialog().notification(__title__, __language__(20001), __icon__, 3000)
        __hyperion__.clear(priority)

    def _clearAll(self):
        xbmcgui.Dialog().notification(__title__, __language__(20001), __icon__, 3000)
        __hyperion__.clearAll()

    def _switch(self):

        if not __addon__.getSetting('switch_type'):
            return

        priority = self.params.get('priority', None)

        if __hyperion__.getState() == 'on':
            __hyperion__.color('black', priority)
        else:
            if __addon__.getSetting('switch_type') == '1':
                __hyperion__.effect(__addon__.getSetting('switch_effect'), priority)
            else:
                __hyperion__.clearAll()

        if __addon__.getSetting('switch_additional'):
            cmd = '%s %s' % (__addon__.getSetting('switch_additional'), __hyperion__.getState())
            log('Script: switch additional %s' % cmd)
            subprocess.Popen(cmd, shell=True)


if __name__ == '__main__':
    log('Script: execute')
    Main()
